extensions/guacamole-vault/modules/guacamole-vault-hv/src/main/java/org/apache/guacamole/vault/hv/secret/HvSecretService.java

```diff
@@ -0,0 +1,335 @@
+ * The configuration used to connect to HV can be set at a global
+ * level using guacamole.properties, or using a connection group
+ * attribute.
+ */
+@Singleton
+public class HvSecretService implements VaultSecretService {
+
+    /**
+     * Logger for this class.
+     */
+    private static final Logger logger = LoggerFactory.getLogger(VaultSecretService.class);
+
+    /**
+     * Service for retrieving configuration information.
+     */
+    @Inject
+    private HvConfigurationService confService;
+
+    /**
+     * Factory for creating HV client instances.
+     */
+    @Inject
+    private HvClientFactory hvClientFactory;
+
+    /**
+     * A map of base-64 encoded JSON HV config blobs to associated HV client instances.
+     * A distinct HV client will exist for every HV config.
+     */
+    private final ConcurrentMap<String, HvClient> hvClientMap = new ConcurrentHashMap<>();
+
+    /**
+     * Create and return a HV client for the provided HV config if not already
+     * present in the client map, otherwise return the existing client entry.
+     *
+     * @param hvConfig
+     *     The base-64 encoded JSON HV config blob associated with the client entry.
+     *     If an associated entry does not already exist, it will be created using
+     *     this configuration.
+     *
+     * @return
+     *     A HV client for the provided HV config if not already present in the
+     *     client map, otherwise the existing client entry.
+     *
+     * @throws GuacamoleException
+     *     If an error occurs while creating the HV client.
+     */
+    private HvClient getClient(@Nonnull String hvConfigBase64)
+            throws GuacamoleException {
+
+        // If a client already exists for the provided config, use it
+        HvClient hvClient = hvClientMap.get(hvConfigBase64);
+        if (hvClient != null)
+            return hvClient;
+
+        // Create and store a new HV client instance for the provided HV config blob
+        Map<String, String> hvConfig = confService.parseHvConfig(hvConfigBase64);
+        hvClient = hvClientFactory.create(hvConfig);
+        HvClient prevClient = hvClientMap.putIfAbsent(hvConfigBase64, hvClient);
+
+        // If the client was already set before this thread got there, use the existing one
+        return prevClient != null ? prevClient : hvClient;
+    }
+
+    @Override
+    public String canonicalize(String nameComponent) {
+        try {
+
+            // As HV notation is essentially a URL, encode all components
+            // using standard URL escaping
+            return URLEncoder.encode(nameComponent, "UTF-8");
+
+        }
+        catch (UnsupportedEncodingException e) {
+            throw new UnsupportedOperationException("Unexpected lack of UTF-8 support.", e);
+        }
+    }
+
+    @Override
+    public Future<String> getValue(
+        UserContext userContext,
+        Connectable connectable,
+        String name
+    ) throws GuacamoleException {
```

Comment:
Please adjust style for these parameters to the established style within the code base.